sys/net/gnrc/routing/rpl/mrhof.c

```diff
@@ -0,0 +1,288 @@
+        return false;
+    }
+
+    return netstats_nb_get(&iface->netif, nce.l2addr, nce.l2addr_len, out);
+}
+
+/**
+ * Select a metric to use as a link metric
+ */
+static inline uint16_t _link_metric(netstats_nb_t *stats)
+{
+    if (stats == NULL) {
+        return MRHOF_MAX_PATH_COST;
+    }
+
+#if IS_USED(MODULE_GNRC_RPL_MRHOF_ETX)
+    return stats->etx;
+#elif IS_USED(MODULE_GNRC_RPL_MRHOF_LQI)
+    /* 255 = best, 0 = worst LQI; map best to ETX = 1, worst to ETX = 7 */
+    return 3 * (0xFF - stats->lqi) + NETSTATS_NB_ETX_DIVISOR;
+#else
+    return MRHOF_MAX_PATH_COST;
+#endif
+}
+
+/**
+ * Retrieve the full path cost of a parent
+ */
+static uint16_t _mrhof_get_path_cost(gnrc_rpl_parent_t *parent, netstats_nb_t *stats)
+{
+    return parent->rank + _link_metric(stats);
+}
+
+/**
+ * Check if a parent is within MAX_LINK_METRIC and MAX_PATH_COST
+ */
+static bool _mrhof_is_acceptable(gnrc_rpl_parent_t *parent, netstats_nb_t *stats)
+{
+    return _link_metric(stats) < MRHOF_MAX_LINK_METRIC &&
+           _mrhof_get_path_cost(parent, stats) < MRHOF_MAX_PATH_COST;
+}
+
+/**
+ * Check if a parent is fresh
+ */
+static bool _mrhof_isfresh(netif_t *netif, netstats_nb_t *stats)
+{
+    /* No statistics is not fresh */
+    if (stats == NULL) {
+        return false;
+    }
+    return netstats_nb_isfresh(netif, stats);
+}
+
+/* Compare acceptability of two parents */
+static int _mrhof_cmp_acceptable(gnrc_rpl_parent_t *p1, netstats_nb_t *p1_stats,
+                                 gnrc_rpl_parent_t *p2, netstats_nb_t *p2_stats)
+{
+    bool p1_acceptable = _mrhof_is_acceptable(p1, p1_stats);
+    bool p2_acceptable = _mrhof_is_acceptable(p2, p2_stats);
+
+    if (p1_acceptable == p2_acceptable) {
+        return 0;
+    }
+
+    /* One of the two parents is not acceptable */
+    if (p1_acceptable == false) {
+        /* p2 is acceptable and p1 not */
+        return 1;
+    }
+
+    if (p2_acceptable == false) {
+        /* p1 is acceptable and p2 not */
+        return -1;
+    }
+
+    return 0;
+}
+
+/* Compare freshness of two parents */
+static int _mrhof_cmp_fresh(netif_t *netif,
+                            netstats_nb_t *p1_stats,
+                            netstats_nb_t *p2_stats)
+{
+    bool p1_isfresh = _mrhof_isfresh(netif, p1_stats);
+    bool p2_isfresh = _mrhof_isfresh(netif, p2_stats);
+
+    if (p1_isfresh == p2_isfresh) {
+        return 0;
+    }
+
+    /* One of the two parents is not acceptable */
+    if (p1_isfresh == false) {
+        /* p2 is acceptable and p1 not */
+        return 1;
+    }
+
+    if (p2_isfresh == false) {
+        /* p1 is acceptable and p2 not */
+        return -1;
+    }
+
+    return 0;
+}
+
+static void reset(gnrc_rpl_dodag_t *dodag)
+{
+    /* Nothing to do in MRHOF */
+    (void) dodag;
+}
+
+/**
+ * Calculate rank additive based on the rank of the parent and the etx to the parent
+ * computed via MAX(pref_parent->rank, 1 + floor(MAX(parents->rank)/MinHopRankIncrease),
```

Comment:
What about

> 3.  The largest calculated Rank among paths through the parent set, minus MaxRankIncrease.

In [RFC 6719, section 3.3: Calculating Rank](https://tools.ietf.org/html/rfc6719#section-3.3):

>    A node sets its Rank to the maximum of three values:
> 
>    1.  The Rank calculated for the path through the preferred parent.
> 
>    2.  The Rank of the member of the parent set with the highest
>        advertised Rank, rounded to the next higher integral Rank, i.e.,
>        to MinHopRankIncrease * (1 + floor(Rank/MinHopRankIncrease)).
> 
>    3.  The largest calculated Rank among paths through the parent set,
>        minus MaxRankIncrease.

Reply:
I think 3 is simply left out based on

> The exact selection of a parent set is an implementation decision.

That's likely what the 

```
/* TODO: consider a parent set of more than 1 parent */
```

was about. But then I'm wondering why the parent set (which is expected to always be one, so there is no `dodag->parents->next`) is being iterated in step 2.

Reply:
From the enumeration of section 3.3:

This is number 1.

https://github.com/RIOT-OS/RIOT/blob/b695b6e4e5bf48f5edd4c1f718940dc3c2c86477/sys/net/gnrc/routing/rpl/mrhof.c#L175-L180

This is number 2.

https://github.com/RIOT-OS/RIOT/blob/b695b6e4e5bf48f5edd4c1f718940dc3c2c86477/sys/net/gnrc/routing/rpl/mrhof.c#L182-L203

This iterates over the parents and take the advertised dag rank and increment it by MinHopRankIncrease

And number 3 is missing as far I know. Number 3 could probably merged into the for loop code of number 2 and take the max of the path through the parent (rank of one parent + etx to that parent).